### PR TITLE
[KERNEL32] LCMapString: Hiragana/Katakana conversion

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Support Japanese Hiragana and Katakana conversion.
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Implement `LCMAP_KATAKANA` and `LCMAP_HIRAGANA` flags.

## TODO

- [x] Do big tests.

### CODE REVIEW

**Comment by katahiromz** [comment]
KVM BEFORE: c762234 https://reactos.org/testman/compare.php?ids=86591
KVM AFTER: 1054882

VBox BEFORE: c762234 https://reactos.org/testman/compare.php?ids=86592
VBox AFTER: 1054882

**Comment by katahiromz** [comment]
KVM BEFORE: c762234 https://reactos.org/testman/compare.php?ids=86591
KVM AFTER: 054d1bf https://reactos.org/testman/compare.php?ids=???
https://reactos.org/testman/compare.php?ids=86591,???

VBox BEFORE: c762234 https://reactos.org/testman/compare.php?ids=86592
VBox AFTER: 054d1bf https://reactos.org/testman/compare.php?ids=???
https://reactos.org/testman/compare.php?ids=86592,???

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_27_03_2023_18_51_24
LGTM.

File: dll/win32/kernel32/winnls/string/lang.c [hunk 1]
```diff
@@ -2008,6 +2008,37 @@ INT WINAPI LCMapStringEx(LPCWSTR name, DWORD flags, LPCWSTR src, INT srclen, LPW
...
+        INT convlen = dst_ptr - dst;
+        for (dst_ptr = dst; convlen; --convlen, ++dst_ptr)
+        {
+            WCHAR wch = *dst_ptr;
+            if (0x30A1 <= wch && wch <= 0x30F3) /* U+30A1 ... U+30F3: Katakana */
+                *dst_ptr = wch - 0x60; /* Katanaka to Hiragana */
+            else if (wch == 0x30F5) /* U+30F5: KATAKANA LETTER SMALL KA */
+                *dst_ptr = 0x3095; /* U+3095: HIRAGANA LETTER SMALL KA */
+            else if (wch == 0x30FD) /* U+30FD: Katakana Iteration Mark */
+                *dst_ptr = 0x309D; /* U+309D: Hiragana Iteration Mark */
```
Comment (HBelusca) [inline dll/win32/kernel32/winnls/string/lang.c:2037]: It seems that for those two characters there is also a subtraction by 0x60 being done.
Couldn't this be grouped as follows? :
```c
else if ((wch == 0x30F5) || wch == 0x30FD)
    *dst_ptr = wch - 0x60;
```
same question also for the other case above, where instead we instead have an addition of 0x60.

**Review by HBelusca — COMMENTED** [review]
Maybe this improvement could be sent to Wine as well.

File: dll/win32/kernel32/winnls/string/lang.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/kernel32/winnls/string/lang.c]: Fixed.

File: dll/win32/kernel32/winnls/string/lang.c [hunk 2]
```diff
@@ -2014,27 +2014,29 @@ INT WINAPI LCMapStringEx(LPCWSTR name, DWORD flags, LPCWSTR src, INT srclen, LPW
         INT convlen = dst_ptr - dst;
         for (dst_ptr = dst; convlen; --convlen, ++dst_ptr)
         {
+            /*
+             * U+3041 ... U+3093: Hiragana
+             * U+3095: HIRAGANA LETTER SMALL KA
```
Comment (HBelusca) [inline dll/win32/kernel32/winnls/string/lang.c:2019]: Why written in all caps by the way? :thinking: :joy:

**Review by HBelusca — APPROVED** [review]

File: dll/win32/kernel32/winnls/string/lang.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/kernel32/winnls/string/lang.c]: Fixed.

**Review by HBelusca — APPROVED** [review]
So, as said previously, it's up to you to send this to Wine if you want.

**Comment by katahiromz** [comment]
> it's up to you to send this to Wine if you want.

Wine has changed its source a lot since our `r52754`. Current Wine is better than our code.

**Pull request merged**